### PR TITLE
patch fixes and test cases

### PR DESCRIPTION
**Pull request by erant10**
This PR addresses the following PATCH issues/cases:
- Apply PATCH to single complex attribute (e.g. User.name)
- Apply PATCH Add to a MultiValue attribute that is missing (should create the list and then add the object)
- Fixes issue where adding an item to a missing attribute create an unmodifiable map
- address the following RFC requirement ([source](https://www.rfc-editor.org/rfc/rfc7644#section-3.5.2)):

  >   For multi-valued attributes, a PATCH operation that sets a value's "primary" sub-attribute to "true" SHALL cause the server to automatically set "primary" to "false" for any other values in the array.

### CODE REVIEW

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java [hunk 1]
```diff
@@ -279,11 +300,14 @@ public <T extends ScimResource> void applyMultiValue(T source, Map<String, Objec
           throw new UnsupportedFilterException("Attribute cannot be added, only comparison expressions are supported when the existing item does not exist.");
         }
         AttributeComparisonExpression comparisonExpression = (AttributeComparisonExpression) valuePathExpression.getAttributeExpression();
+        checkPrimary(subAttributeName, items, value);
 
-      items.add(Map.of(
+        items.add(new HashMap<>(Map.of(
```
Comment (bdemers) [inline scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java:308]: nit: Does this need to be a HashMap?

File: scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java [hunk 2]
```diff
@@ -127,6 +127,57 @@ public void applyAddToMultiValuedComplexAttribute() {
...
+        .setType("work")
+        .setPostalCode("ko4 8qq"));
+    assertThat(updatedUser.getAddresses()).isEqualTo(expectedAddresses);
+  }
+
+  @Test
+  public void settingPrimaryOnMultiValuedShouldResetAllOthersToFalse() {
+    // https://www.rfc-editor.org/rfc/rfc7644#section-3.5.2
+    ScimUser user =  user();
+    PatchOperation op = patchOperation(REPLACE, "emails[type eq \"home\"].primary", true);
```
Comment (bdemers) [inline scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java:167]: This is a great catch I hadn't noticed that in the spec!

**Review by bdemers — APPROVED** [review]
Great change!!

One minor question before we merge

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java (hunk 1 above)
Comment (erant10) [reply scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java:308]: No I suppose it doesn't. Thing is `Map.of` creates an Immutable map by default, which later causes trouble when we try to call `items.put`... I was trying to solve it when I came across [this](https://stackoverflow.com/questions/54963337/create-mutable-collections-map-of-and-list-of)....
But I am open to any suggestions of course

File: scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java (hunk 2 above)
Comment (erant10) [reply scim-core/src/test/java/org/apache/directory/scim/core/repository/PatchHandlerTest.java:167]: Neither did I lol
I'll pass the message on to my teammate @photonicworld

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java [hunk 3]
```diff
@@ -171,6 +171,18 @@ private static void checkMutability(Attribute attribute, Object currentValue) th
     }
   }
 
+  private static void checkPrimary(String subAttributeName, Collection<Map<String, Object>> items, Object value)
+  {
+    if (subAttributeName.equals("primary") && value.equals(true)) {
```
Comment (erant10) [inline scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java]: @bdemers where would be a good place to declare `static final PRIMARY = "primary"`? 
`Schema` maybe?

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java (hunk 1 above)
Comment (bdemers) [reply scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java:308]:
```suggestion
        // Add a new mutable map
        items.add(new HashMap<>(Map.of(
```

File: scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java (hunk 3 above)
Comment (bdemers) [reply scim-core/src/main/java/org/apache/directory/scim/core/repository/PatchHandlerImpl.java]: 🤔 for now, just stick it in `PatchHandlerImpl`, The special rules for `primary` seems to only apply to PATCH operations.

**Comment by bdemers** [comment]
@erant10 I found (and fixed) a flaky test that was causing this PR (and a few others to fail).
Can you rebase this PR?

**Comment by erant10** [comment]
@bdemers rebased. Thanks for fixing the test :)

**Review by bdemers — APPROVED** [review]

**Comment by bdemers** [comment]
Thanks again @erant10!

**Pull request merged**